lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +78,522 @@ struct DiagRestore {
+                                       NumOutputVectors);
+    if (PSVTable != DxilTable)
+      return true;
+  }
+
+  if (UsesViewID && NumOutputVectors > 0) {
+    PSVComponentMask DxilMask(ArrayMask.data(), NumOutputVectors);
+    if (PSVMask != DxilMask)
+      return true;
+  }
+  return false;
+}
+
+void PSVContentVerifier::VerifyViewIDDependence(PSVRuntimeInfo1 *PSV1) {
+  SimpleViewIDState ViewIDState(DM.GetSerializedViewIdState(),
+                                PSV.GetShaderKind(), PSV1->UsesViewID);
+
+  if (!ViewIDState.IsValid) {
+    EmitInvalidError("ViewIDState");
+    return;
+  }
+
+  unsigned NumInputScalars = ViewIDState.NumInputSigScalars;
+  auto &NumOutputScalars = ViewIDState.NumOutputSigScalars;
+  unsigned NumPCOrPrimScalars = ViewIDState.NumPCOrPrimSigScalars;
+  bool Mismatched = false;
+  for (unsigned i = 0; i < DXIL::kNumOutputStreams; i++) {
+    // NumInputScalars as input, NumOutputScalars[i] as output.
+    uint32_t NumInputVectors = llvm::RoundUpToAlignment(NumInputScalars, 4) / 4;
+    uint32_t NumOutputVectors =
+        llvm::RoundUpToAlignment(NumOutputScalars[i], 4) / 4;
+    Mismatched |= ViewIDTableAndMaskMismatched(
+        PSV.GetInputToOutputTable(i), ViewIDState.InputToOutputTable[i],
+        PSV.GetViewIDOutputMask(i), ViewIDState.ViewIDOutputMask[i],
+        NumInputVectors, NumOutputVectors, PSV1->UsesViewID);
+  }
+
+  // NumInputScalars as input, NumPCOrPrimScalars as output.
+  if (PSV1->ShaderStage == static_cast<uint8_t>(PSVShaderKind::Hull) ||
+      PSV1->ShaderStage == static_cast<uint8_t>(PSVShaderKind::Mesh)) {
+    uint32_t NumInputVectors = llvm::RoundUpToAlignment(NumInputScalars, 4) / 4;
+    uint32_t NumOutputVectors =
+        llvm::RoundUpToAlignment(NumPCOrPrimScalars, 4) / 4;
+    Mismatched |= ViewIDTableAndMaskMismatched(
+        PSV.GetInputToPCOutputTable(), ViewIDState.InputToPCOutputTable,
+        PSV.GetViewIDPCOutputMask(), ViewIDState.ViewIDPCOutputMask,
+        NumInputVectors, NumOutputVectors, PSV1->UsesViewID);
+  }
+
+  // NumPCOrPrimScalars as input, NumOutputScalars[0] as output.
+  if (PSV1->ShaderStage == static_cast<uint8_t>(PSVShaderKind::Domain)) {
+    uint32_t NumInputVectors =
+        llvm::RoundUpToAlignment(NumPCOrPrimScalars, 4) / 4;
+    uint32_t NumOutputVectors =
+        llvm::RoundUpToAlignment(NumOutputScalars[0], 4) / 4;
+    Mismatched |= ViewIDTableAndMaskMismatched(
+        PSV.GetPCInputToOutputTable(), ViewIDState.PCInputToOutputTable,
+        PSVComponentMask(), ViewIDState.ViewIDOutputMask[0], NumInputVectors,
+        NumOutputVectors, false);
+  }
+  if (Mismatched) {
+    std::string Str;
+    raw_string_ostream OS(Str);
+    PSV.PrintViewIDState(OS);
+    OS.flush();
+    std::string Str1;
+    raw_string_ostream OS1(Str1);
+    ViewIDState.Print(OS1, static_cast<PSVShaderKind>(PSV1->ShaderStage),
+                      PSV1->UsesViewID);
+    OS1.flush();
+    EmitMismatchError("ViewIDState", Str, Str1);
```

Comment:
Feels a bit weird to submit multiline strings as parameters to an error message, it would be easier to read if the specific mismatch were identified and printed. I'm not sure I'm ready to insist on that though.

Reply:
For the error message, my feeling is more information is better because the input of validation is a container which is not easy to read. Having only a single piece of the message makes it difficult to understand the full scope of the problem.

Reply:
I don't agree, but I don't think it's worth holding this up. Hopefully these errors will be seen infrequently if at all.